desktop/libs/notebook/src/notebook/connectors/trino.py

```diff
@@ -0,0 +1,306 @@
+#!/usr/bin/env python
+# Licensed to Cloudera, Inc. under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  Cloudera, Inc. licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import logging
+import json
+import posixpath
+import sys
+import textwrap
+
+from desktop.lib.i18n import force_unicode
+from desktop.lib.rest.http_client import HttpClient, RestException
+from desktop.lib.rest.resource import Resource
+from urllib.parse import urlparse
+import urllib.parse
+from notebook.connectors.base import Api, QueryError
+
+
+from django.utils.translation import gettext as _
+from notebook.connectors.base import Api, QueryError
+
+from trino.client import ClientSession, TrinoRequest, TrinoQuery
+
+def query_error_handler(func):
+  def decorator(*args, **kwargs):
+    try:
+      return func(*args, **kwargs)
+    except RestException as e:
+      try:
+        message = force_unicode(json.loads(e.message)['errors'])
+      except:
+        message = e.message
+      message = force_unicode(message)
+      raise QueryError(message)
+    except Exception as e:
+      message = force_unicode(str(e))
+      raise QueryError(message)
+  return decorator
+
+
+
+class TrinoApi(Api):
+
+  def __init__(self, user, interpreter=None):
+    Api.__init__(self, user, interpreter=interpreter)
+
+    self.options = interpreter['options']
+    
+    api_url = self.options['url']
+    hostname, port = self.get_hostname_and_port(api_url)
+    trino_session = self.get_trino_client_session(api_url)
+    
+    self.db = TrinoRequest(hostname, port, trino_session)
+
+
+  def get_hostname_and_port(self, api_url):
+    parsed_url = urlparse(api_url)
+    hostname = parsed_url.hostname
+    port = parsed_url.port
+    return hostname, port
+
+
+  @query_error_handler
+  def create_session(self, lang=None, properties=None):
+    pass
+
+
+  def get_trino_client_session(self, url):
+    catalog = urlparse(url).path.split('/')[-1]
+    user = self.user.username
+
+    return ClientSession(user, catalog)
+
+
+  @query_error_handler
+  def execute(self, notebook, snippet):
+    
+    statement = snippet['statement'].strip().rstrip(';')
```

Comment:
Does it throw error its there a ending ;?

Reply:
meaning?